### PR TITLE
Nimble Newts

### PR DESCRIPTION
**Pull request by AvianAnalyst**

### CODE REVIEW

**Comment by SebastiaanZ** [comment]
No merge issues

No flake8 linting errors

File: nimble-newts/askgrieves/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,9 @@
+FROM python:3.8.5-buster
```
Comment (jb3) [inline nimble-newts/askgrieves/Dockerfile:1]: Good use of Docker!

If you copy the Pipfile & lockfile separately you can make use of Docker cache to speed up build times. The ideal Dockerfile would go like:

- Install `pipenv`
- Copy `Pipfile` and `Pipfile.lock` to the directory
- Install dependencies
- Run all other commands

This means that pipenv does not have to be reinstalled if you change dependencies because it was built in an earlier layer where dependencies were irrelevant.

File: nimble-newts/askgrieves/askgrieves/settings.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = 'jc&@p_n-q^e4z&-l0y@jo$n+bg0y4v^vvvx=yoqjkyf-ugc+_2'
```
Comment (jb3) [inline nimble-newts/askgrieves/askgrieves/settings.py:23]: These should be set to something like an environment variable set in production.

This is the secret that Django admin will use and keeping the secret key open could allow for tampering.

File: nimble-newts/README.md [hunk 3]
```diff
@@ -1,2 +1,18 @@
 # Nimble Newts
-This is the default README of your team's project. Please replace this by a README with more information on your project. At the very least, your README should contain information on how to set-up and run your project.
+
```
Comment (jb3) [inline nimble-newts/README.md:2]: Screenshots in READMEs are always great, it would be nice to have one here.

File: nimble-newts/README.md [hunk 4]
```diff
@@ -1,2 +1,18 @@
...
+A simple bot that responds to requests for Wikipedia articles.
+
+## How to run:
```
Comment (jb3) [inline nimble-newts/README.md:5]: This how to run is pretty vague, it could do with a bit more of a description about what the project is (and again, screenshots!)

File: nimble-newts/askgrieves/chatbot/models.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from django.db import models
+
+
+class Chatbot(models.Model):
+    name = models.CharField(max_length=500)
+
+    def __str__(self):
+        return f"name={self.name}"
```
Comment (jb3) [inline nimble-newts/askgrieves/chatbot/models.py:8]: This looks like somewhat of an attempt at a repr implementation, I can't see `name=abc` being useful otherwise.

Maybe this should be implemented as a repr similar to:

```py
In [4]: class Chatbot:
   ...:     def __init__(self, name):
   ...:         self.name = name
   ...:     def __repr__(self):
   ...:         return f"Chatbot(name={repr(self.name)})"
   ...:
   ...:

In [5]: Chatbot("abc")
Out[5]: Chatbot(name='abc')
```

File: nimble-newts/askgrieves/chatbot/models.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class WikiArticle(models.Model):
+    name = models.CharField(max_length=500)
+    summary = models.TextField(max_length=100)
+    full_page = models.TextField(max_length=500)
+
+    def __str__(self):
+        return f"name={self.name} summary={self.summary}"
```
Comment (jb3) [inline nimble-newts/askgrieves/chatbot/models.py:17]: Similar to the comment on the Chatbot model, this isn't a helpful output.

File: nimble-newts/askgrieves/chatbot/templates/home.html [hunk 7]
```diff
@@ -0,0 +1,30 @@
+<!DOCTYPE html>
+<html lang="en">
+    <head>
+        {% load message_filter %}
+        {% load static %}
+        <link rel="stylesheet" href="{% static 'css/chat.css' %}" type="text/css" />
+        <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css" integrity="sha384-JcKb8q3iqJ61gNV9KGb8thSsNjpSL0n8PARn9HuZOnIxN0hoP+VmmDGMN5t9UJ0Z" crossorigin="anonymous">
```
Comment (jb3) [inline nimble-newts/askgrieves/chatbot/templates/home.html:7]: Good use of integrity checking!

File: nimble-newts/askgrieves/chatbot/tests/test_getwiki.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+from django.http import JsonResponse
+
+import pytest
+
+import json
+
+
+@pytest.mark.django_db
+class TestWiki:
```
Comment (jb3) [inline nimble-newts/askgrieves/chatbot/tests/test_getwiki.py:9]: Tests! This is awesome, great work!

File: nimble-newts/askgrieves/chatbot/views.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+class Message:
+    send = True
+
+    def __init__(self, text):
+        if Message.send:
+            self.action = 'send'
+        else:
+            self.action = 'receive'
+        Message.send = not self.send
+        self.text = text
```
Comment (jb3) [inline nimble-newts/askgrieves/chatbot/views.py:21]: The usage of attributes here is quite odd and hard to follow.

File: nimble-newts/askgrieves/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
+version: "2.0"
```
Comment (jb3) [inline nimble-newts/askgrieves/docker-compose.yml:1]: Good usage of Docker compose!

**Review by jb3 — COMMENTED** [review]
Great work Nimble Newts!

This is a really nice looking Django application.

Your usage of Docker and Docker Compose leads to a very easy set up process. However, the howto section and the rest of your README could do with some extra work to include things like screenshots.

Code quality is good, you have no linting errors but you could do with more comments and docstrings, there are barely any across the project. Your usage of pytest for your application is great.

As mentioned in settings.py, in production you should not be leaving DEBUG to True and SECRET_KEY to a literal in production as it could allow malicious users to tamper with your database if they get access.

Overall this is good work with good code quality and a good structure. Congratulations!
